File: app/src/main/java/dev/ragnarok/fenrir/view/FadeAnimDrawable.java
package dev.ragnarok.fenrir.view;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.SystemClock;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.NonNull;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.api.model.PlayerCoverBackgroundSettings;
import dev.ragnarok.fenrir.settings.Settings;

public final class FadeAnimDrawable extends BitmapDrawable implements Animatable {
    // Only accessed from main thread.
    private static final float FADE_DURATION = 400f; //ms
    private final long startTimeMillis;
    private final PlayerCoverBackgroundSettings settings;
    private boolean animatingFade;
    private boolean animating;
    private int alpha = 0xFF;
    private View currentParentView;
    private float targetRotation;

    public FadeAnimDrawable(Context context, @NonNull Bitmap bitmap) {
        super(context.getResources(), bitmap);
        animatingFade = true;
        animating = false;
        startTimeMillis = SystemClock.uptimeMillis();
        settings = Settings.get().other().getPlayerCoverBackgroundSettings();
        targetRotation = 0;
    }

    public static void setBitmap(@NonNull ImageView target, Context context, @NonNull Bitmap bitmap, boolean start) {
        PlayerCoverBackgroundSettings settings = Settings.get().other().getPlayerCoverBackgroundSettings();
        if (!settings.enabled_rotation || settings.rotation_speed <= 0) {
            FadeDrawable.setBitmap(target, context, bitmap);
            return;
        }
        Drawable placeholder = target.getDrawable();
        if (placeholder instanceof Animatable) {
            ((Animatable) placeholder).stop();
        }
        FadeAnimDrawable drawable = new FadeAnimDrawable(context, bitmap);
        drawable.setCurrentParentView(target);
        if (start) {
            drawable.start();
        }
        target.setImageDrawable(drawable);
    }

    public void setCurrentParentView(View view) {
        currentParentView = view;
    }

    protected void invalidateInternal() {
        if (currentParentView != null) {
            currentParentView.invalidate();
        }
        if (getCallback() != null) {
            invalidateSelf();
        }
    }

    @Override
    public void draw(Canvas canvas) {
        canvas.scale(settings.zoom, settings.zoom, (float) getBitmap().getWidth() / 2, (float) getBitmap().getHeight() / 2);
        if (!animatingFade) {
            if (animating) {
                if (!settings.invert_rotation) {
                    targetRotation += settings.rotation_speed;
                    if (targetRotation > 360) {
                        targetRotation = 0;
                    }
                } else {
                    targetRotation -= settings.rotation_speed;
                    if (targetRotation < -360) {
                        targetRotation = 0;
                    }
                }

                canvas.rotate((float) Math.toDegrees(targetRotation), (float) getBitmap().getWidth() / 2, (float) getBitmap().getHeight() / 2);
                try {
                    super.draw(canvas);
                    invalidateInternal();
                } catch (Exception e) {
                    if (Constants.IS_DEBUG) {
                        e.printStackTrace();
                    }
                }
            } else {
                canvas.rotate((float) Math.toDegrees(targetRotation), (float) getBitmap().getWidth() / 2, (float) getBitmap().getHeight() / 2);
                super.draw(canvas);
            }
        } else {
            float normalized = (SystemClock.uptimeMillis() - startTimeMillis) / FADE_DURATION;
            if (normalized >= 1f) {
                animatingFade = false;
                try {
                    super.draw(canvas);
                    if (animating) {
                        invalidateInternal();
                    }
                } catch (Exception e) {
                    if (Constants.IS_DEBUG) {
                        e.printStackTrace();
                    }
                }
            } else {
                // setAlpha will call invalidateSelf and drive the animation.
                int partialAlpha = (int) (alpha * normalized);
                super.setAlpha(partialAlpha);
                try {
                    super.draw(canvas);
                } catch (Exception e) {
                    if (Constants.IS_DEBUG) {
                        e.printStackTrace();
                    }
                }
                super.setAlpha(alpha);
            }
        }
    }

    @Override
    public void setAlpha(int alpha) {
        this.alpha = alpha;
        super.setAlpha(alpha);
    }

    @Override
    public void start() {
        if (!animating) {
            animating = true;
            invalidateInternal();
        }
    }

    @Override
    public void stop() {
        if (animating) {
            animating = false;
            invalidateInternal();
        }
    }

    @Override
    public boolean isRunning() {
        return animating;
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/view/media/FadeAnimDrawable.kt
package dev.ragnarok.fenrir.view.media

import android.graphics.*
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.os.SystemClock
import android.view.View
import androidx.annotation.ColorInt
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.api.model.PlayerCoverBackgroundSettings
import dev.ragnarok.fenrir.settings.Settings
import kotlin.math.abs
import kotlin.math.cos


class FadeAnimDrawable(private val bitmap: Bitmap, @ColorInt private val surface: Int) : Drawable(),
    Animatable {
    private val startTimeMillis: Long = SystemClock.uptimeMillis()
    private val settings: PlayerCoverBackgroundSettings =
        Settings.get().other().playerCoverBackgroundSettings
    private var animatingFade = true
    private var animating = false
    private var pAlpha = 0xCC
    private var targetRotation: Float = 0f
    private var targetSaturation = 1f
    private val paint =
        Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG)
    private val rBmp = Rect()
    private val rDest = Rect()
    private var lastFrame: Long = 0
    private var rectFirst: Rect = Rect()
    private var rectSec: Rect = Rect()
    private val colorMatrix = ColorMatrix()
    private lateinit var cFilter: ColorMatrixColorFilter

    @ColorInt
    private val surfaceColor =
        Color.argb(255, Color.red(surface), Color.green(surface), Color.blue(surface))
    private val surfaceNoAlpha =
        Color.argb(0, Color.red(surface), Color.green(surface), Color.blue(surface))

    private fun invalidateInternal() {
        if (callback != null) {
            invalidateSelf()
        }
    }

    private fun printBitmap(canvas: Canvas) {
        val pMaxDst = (bounds.width()
            .coerceAtLeast(bounds.height()) * (if (settings.enabled_rotation && settings.rotation_speed > 0) settings.zoom else 1f)).toInt()
        rDest.left = bounds.left - (pMaxDst - bounds.width()) / 2
        rDest.right = bounds.right + (pMaxDst - bounds.width()) / 2
        rDest.top = bounds.top - (pMaxDst - bounds.height()) / 2
        rDest.bottom = bounds.bottom + (pMaxDst - bounds.height()) / 2
        canvas.save()
        canvas.rotate(
            targetRotation, bounds.left + bounds.width().toFloat() / 2, bounds.height()
                .toFloat() / 2
        )
        canvas.drawBitmap(bitmap, rBmp, rDest, paint)
        canvas.restore()
    }

    private fun clamp(min: Float, max: Float, value: Float): Float {
        if (value < min) {
            return min
        } else if (value > max) {
            return max
        }
        return value
    }

    private fun doRotate(canvas: Canvas): Boolean {
        var invalidate = false
        if (!animatingFade) {
            if (animating && settings.enabled_rotation && settings.rotation_speed > 0) {
                if (settings.invert_rotation) {
                    targetRotation += settings.rotation_speed * clamp(
                        0.2f,
                        3f,
                        (System.currentTimeMillis() - lastFrame) / 17f
                    )
                    if (targetRotation > 360) {
                        targetRotation = 0f
                    }
                } else {
                    targetRotation -= settings.rotation_speed * clamp(
                        0.2f,
                        3f,
                        (System.currentTimeMillis() - lastFrame) / 17f
                    )
                    if (targetRotation < -360) {
                        targetRotation = 0f
                    }
                }
                if (settings.fade_saturation) {
                    targetSaturation =
                        clamp(
                            0.2f,
                            1f,
                            abs(cos(Math.toRadians(targetRotation.toDouble())).toFloat())
                        )
                    colorMatrix.setSaturation(targetSaturation)
                    cFilter = ColorMatrixColorFilter(colorMatrix)
                    paint.colorFilter = cFilter
                }

                lastFrame = System.currentTimeMillis()
            }
            try {
                printBitmap(canvas)
                invalidate = true
            } catch (e: Exception) {
                if (Constants.IS_DEBUG) {
                    e.printStackTrace()
                }
            }
        } else {
            val normalized = (SystemClock.uptimeMillis() - startTimeMillis) / FADE_DURATION
            if (normalized >= 1f) {
                animatingFade = false
                try {
                    printBitmap(canvas)
                    if (animating && settings.enabled_rotation && settings.rotation_speed > 0) {
                        invalidate = true
                    }
                } catch (e: Exception) {
                    if (Constants.IS_DEBUG) {
                        e.printStackTrace()
                    }
                }
            } else {
                // setAlpha will call invalidateSelf and drive the animation.
                val partialAlpha = (pAlpha * normalized).toInt()
                paint.alpha = partialAlpha
                try {
                    printBitmap(canvas)
                } catch (e: Exception) {
                    if (Constants.IS_DEBUG) {
                        e.printStackTrace()
                    }
                }
                paint.alpha = pAlpha
                invalidate = true
            }
        }
        return invalidate
    }

    override fun draw(canvas: Canvas) {
        val invalidate = doRotate(canvas)

        rectFirst.left = bounds.left
        rectFirst.right = bounds.right
        rectFirst.top = bounds.top
        rectFirst.bottom = (bounds.bottom * 0.1f).toInt()

        rectSec.left = bounds.left
        rectSec.right = bounds.right
        rectSec.top = bounds.bottom - (bounds.bottom * 0.5f).toInt()
        rectSec.bottom = bounds.bottom

        paint.shader = LinearGradient(
            rectFirst.left.toFloat(),
            rectFirst.top.toFloat(),
            rectFirst.left.toFloat(),
            rectFirst.bottom.toFloat(),
            surfaceColor,
            surfaceNoAlpha,
            Shader.TileMode.CLAMP
        )
        canvas.drawRect(rectFirst, paint)

        paint.shader = LinearGradient(
            rectSec.left.toFloat(),
            rectSec.top.toFloat(),
            rectSec.left.toFloat(),
            rectSec.bottom.toFloat(),
            surfaceNoAlpha,
            surfaceColor,
            Shader.TileMode.CLAMP
        )
        canvas.drawRect(rectSec, paint)

        if (invalidate) {
            invalidateInternal()
        }
    }

    override fun setAlpha(alpha: Int) {
        this.pAlpha = alpha
        invalidateInternal()
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        throw UnsupportedOperationException("setColorFilter unsupported!")
    }

    override fun getOpacity(): Int {
        return PixelFormat.TRANSLUCENT
    }

    override fun start() {
        if (!animating) {
            animating = true
            lastFrame = System.currentTimeMillis()
            invalidateInternal()
        }
    }

    override fun stop() {
        if (animating) {
            animating = false
            invalidateInternal()
        }
    }

    override fun isRunning(): Boolean {
        return animating
    }

    companion object {
        // Only accessed from main thread.
        private const val FADE_DURATION = 400f //ms
        fun setBitmap(target: View, bitmap: Bitmap, start: Boolean, @ColorInt surface: Int) {
            val settings = Settings.get().other().playerCoverBackgroundSettings
            val drawable = FadeAnimDrawable(bitmap, surface)
            drawable.callback = target
            if (start && settings.enabled_rotation && settings.rotation_speed > 0) {
                drawable.start()
            }
            target.background = drawable
        }
    }

    init {
        val pMin = bitmap.height.coerceAtMost(bitmap.width)
        rBmp.left = (bitmap.width - pMin) / 2
        rBmp.top = (bitmap.height - pMin) / 2
        rBmp.right = pMin - rBmp.left
        rBmp.bottom = pMin - rBmp.top
    }
}